File: Javascript/Traversal/Keys-And-Rooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
const canVisitAllRooms = (rooms) => {
  let keys = [0];
  let visited = new Set();

  while (keys.length > 0) {
    const room = keys.pop();

    if (!visited.has(room)) {
      visited.add(room);
      keys.push(...rooms[room]);
    }
  }
  if (visited.size == rooms.length) return true;

  return false;
};

File: Rust/Traversal/Keys-And-Rooms.rs
struct Solution;

impl Solution {
    pub fn can_visit_all_rooms(rooms: Vec<Vec<i32>>) -> bool {
        let mut count = 1;
        let mut keys: Vec<usize> = vec![0];
        let mut visited: Vec<bool> = vec![false; rooms.len()];

        visited[0] = true;

        while let Some(room) = keys.pop() {
            for key in rooms[room].iter().map(|key| *key as usize) {
                if !visited[key] {
                    count += 1;
                    if count == rooms.len() {
                        return true;
                    }
                    visited[key] = true;
                    keys.push(key);
                }
            }
        }
        false
    }
}
fn main() {}
